nuclia/sdk/kb.py

```diff
@@ -167,6 +171,109 @@ def notifications(self, **kwargs):
         for notification in response.iter_lines():
             yield Notification.model_validate_json(notification)
 
+    @kb
+    def copy(
+        self,
+        *,
+        rid: Optional[str] = None,
+        slug: Optional[str] = None,
+        destination: str,
+        **kwargs,
+    ):
+        ndb = kwargs["ndb"]
+        if rid:
+            res = ndb.ndb.get_resource_by_id(
+                kbid=ndb.kbid,
+                rid=rid,
+                query_params={
+                    "show": ["basic", "origin", "extra", "values", "security"]
+                },
+            )
+        elif slug:
+            res = ndb.ndb.get_resource_by_slug(
+                kbid=ndb.kbid,
+                slug=slug,
+                query_params={
+                    "show": ["basic", "origin", "extra", "values", "security"]
+                },
+            )
+        else:
+            raise ValueError("Either rid or slug must be provided")
+        data = {
+            "kbid": destination,
+            "slug": res.slug,
+            "title": res.title,
+            "summary": res.summary,
+            "icon": res.icon,
+            "origin": res.origin,
+            "extra": res.extra,
+            "usermetadata": res.usermetadata,
+            "fieldmetadata": res.fieldmetadata,
+            "security": res.security,
+        }
+        files_to_upload = []
+        if res.data.conversations:
+            data["conversations"] = dict(
+                zip(
+                    res.data.conversations.keys(),
+                    [dict(v.value) for v in res.data.conversations.values()],
+                )
+            )
+        if res.data.links:
+            data["links"] = dict(
+                zip(
+                    res.data.links.keys(),
+                    [dict(v.value) for v in res.data.links.values()],
+                )
+            )
+        if res.data.texts:
+            data["texts"] = dict(
+                zip(
+                    res.data.texts.keys(),
+                    [dict(v.value) for v in res.data.texts.values()],
+                )
+            )
+        if res.data.files:
+            remote_files = {}
+            for file_id, file in res.data.files.items():
+                if file.value.external:
+                    remote_files[file_id] = file.value
+                else:
+                    files_to_upload.append({"id": file_id, "data": file.value})
+        destination_kb = get_client(destination)
+        uuid = self.resource.create(ndb=destination_kb, **data)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            for file_data in files_to_upload:
+                file_path = os.path.join(tmpdirname, file_data["data"].file.filename)
+                self.resource.download_file(
+                    rid=res.id, file_id=file_data["id"], output=file_path, **kwargs
+                )
+                self.upload.file(path=file_path, ndb=destination_kb, rid=uuid)
```

Comment:
This is downloading all the binary into a temporary file, then uploading it again to nuclia and finally removing the temp file.

It's simple and works fine, but if it needs to be speed up in the future, you can consider chaining the download stream response with the upload request.

Reply:
There is already thius feature in the https://github.com/nuclia/nuclia.py/blob/main/nuclia/sdk/upload.py#L239 for sync and async